File: src/connectors/shodan_internetdb.js
import { BaseConnector } from './base.js';
import https from 'https';

export class ShodanInternetDBConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'Shodan InternetDB';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const data = await this._fetch(`https://internetdb.shodan.io/${encodeURIComponent(inputValue)}`);
            const entities = [];

            if (data) {
                if (data.ports?.length) {
                    for (const port of data.ports) {
                        entities.push({
                            type: 'port',
                            data: { number: port, ip: inputValue, source: 'shodan_internetdb' },
                            confidence: 0.9
                        });
                    }
                }

                if (data.hostnames?.length) {
                    for (const hostname of data.hostnames) {
                        entities.push({
                            type: 'domain',
                            data: { name: hostname, source: 'shodan_internetdb' },
                            confidence: 0.8,
                            asKnown: { type: 'domain', value: hostname }
                        });
                    }
                }

                if (data.vulns?.length) {
                    entities.push({
                        type: 'ip',
                        data: {
                            address: inputValue,
                            vulns: data.vulns,
                            ports: data.ports,
                            cpes: data.cpes,
                            source: 'shodan_internetdb'
                        },
                        confidence: 0.9
                    });
                }

                if (data.tags?.length) {
                    entities.push({
                        type: 'ip',
                        data: { address: inputValue, tags: data.tags, source: 'shodan_internetdb_tags' },
                        confidence: 0.8
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _fetch(url) {
        return new Promise((resolve, reject) => {
            const req = https.get(url, { timeout: 10000 }, res => {
                if (res.statusCode === 404) { res.resume(); return resolve(null); }
                if (res.statusCode >= 400) { res.resume(); return reject(new Error(`InternetDB HTTP ${res.statusCode}`)); }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve(null); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('InternetDB timeout')); });
        });
    }

    async healthCheck() {
        return { ok: true, version: 'free', hasApiKey: true };
    }
}

File: cyclops/connectors/shodan_internetdb.py
import aiohttp
from cyclops.connectors.base import BaseConnector


class ShodanInternetDBConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "Shodan InternetDB"

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            async with aiohttp.ClientSession() as session:
                async with session.get(f"https://internetdb.shodan.io/{input_value}", timeout=aiohttp.ClientTimeout(total=10)) as resp:
                    if resp.status == 404:
                        data = None
                    elif resp.status >= 400:
                        raise Exception(f"InternetDB HTTP {resp.status}")
                    else:
                        data = await resp.json()
            if data:
                for port in data.get("ports", []):
                    entities.append({"type": "port", "data": {"number": port, "ip": input_value, "source": "shodan_internetdb"}, "confidence": 0.9})
                for hostname in data.get("hostnames", []):
                    entities.append({"type": "domain", "data": {"name": hostname, "source": "shodan_internetdb"}, "confidence": 0.8, "as_known": {"type": "domain", "value": hostname}})
                if data.get("vulns"):
                    entities.append({"type": "ip", "data": {"address": input_value, "vulns": data["vulns"], "ports": data.get("ports"), "cpes": data.get("cpes"), "source": "shodan_internetdb"}, "confidence": 0.9})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "version": "free"}
